File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/DistinctEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import org.opencds.cqf.cql.engine.execution.Context;

/*
distinct(argument List<T>) List<T>

The distinct operator returns the given list with duplicates eliminated using equality semantics.

If the argument is null, the result is null.
*/

public class DistinctEvaluator extends org.cqframework.cql.elm.execution.Distinct
{

    public static List<Object> distinct(Iterable<?> source, Context context)
    {
        if (source == null)
        {
            return null;
        }

        List<Object> result = new ArrayList<>();
        for (Object element : source)
        {
            if (element == null && result.parallelStream().noneMatch(Objects::isNull))
            {
                result.add(null);
                continue;
            }

            Object in = InEvaluator.in(element, result, null, context);

            if (in == null) continue;

            if (!(Boolean) in) result.add(element);
        }

        return result;
    }

    @Override
    protected Object internalEvaluate(Context context)
    {
        Object value = this.getOperand().evaluate(context);
        return distinct((Iterable<?>)value, context);
    }
}

File: engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/DistinctEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlList

/*
distinct(argument List<T>) List<T>

The distinct operator returns the given list with duplicates eliminated using equality semantics.

If the argument is null, the result is null.
*/
object DistinctEvaluator {
    fun distinct(source: Value?, state: State?): List? {
        if (source == null) {
            return null
        }

        if (source is List) {
            val result = mutableListOf<Value?>()
            for (element in source) {
                if (element == null && result.none { obj -> obj == null }) {
                    result.add(null)
                    continue
                }

                val `in` = InEvaluator.`in`(element, result.toCqlList(), null, state) ?: continue

                if (!`in`.value) result.add(element)
            }

            return result.toCqlList()
        }

        throw InvalidOperatorArgument("distinct(List<T>)", "distinct(${source.typeAsString})")
    }
}
